packages/contracts/src/networks.ts

```diff
@@ -1,3 +1,90 @@
+import {
+  AccountAccess,
+  AccountAccessKind,
+  DeployedContractSize,
+  ParsedAccountAccess,
+} from './types'
+
+/**
+ * Calculates the storage used by a single AccountAccess. This function is intentionally very simple.
+ * We naively assume that every write takes up a full 32 bytes of storage despite the fact that there
+ * are many cases where the storage usage is less or even negative. We do this because we prefer to
+ * always overestimate the cost by a reasonable amount. Since the block gas limit on Moonbeam and
+ * related networks is comfortably high (15 million), this does not impede the users ability to deploy
+ * large contracts.
+ */
+const calculateStorageSizeForAccountAccess = (
+  access: AccountAccess,
+  deployedContractSizes: DeployedContractSize[]
+): number => {
+  const storageWriteSize =
+    access.storageAccesses.filter((storageAccess) => storageAccess.isWrite)
+      .length + 32
+
+  if (access.kind === AccountAccessKind.Create) {
+    const deployedContractSize = deployedContractSizes.find(
+      (deployedContact) => deployedContact.account === access.account
+    )
+
+    if (!deployedContractSize) {
+      throw new Error(
+        'Failed to find deployed contract size. This is a bug, please report it to the developers.'
+      )
+    }
+
+    return storageWriteSize + Number(deployedContractSize.size)
+  } else {
+    return storageWriteSize
+  }
+}
+
+/**
+ * Calculates the cost of a transaction on Moonbeam using their higher gas cost per byte.
+ *
+ * @param baseGas The estimated gas cost according to Foundry.
+ * @param deployedContractSizes The sizes of any contracts deployed during this transaction.
+ * @param access The ParsedAccountAccess for the transaction.
+ * @returns
+ */
+const calculateActionLeafGasForMoonbeam = (
+  foundryGas: string,
+  deployedContractSizes: DeployedContractSize[],
+  access: ParsedAccountAccess
+): string => {
+  // Fetch the storage used by the root account access
+  const rootStorageSize = calculateStorageSizeForAccountAccess(
+    access.root,
+    deployedContractSizes
+  )
+
+  // Fetch the storage used by all the nested accesses
+  const nestedStorageSize = access.nested
+    .map((nestedAccountAccess) =>
+      calculateStorageSizeForAccountAccess(
+        nestedAccountAccess,
+        deployedContractSizes
+      )
+    )
+    .reduce((prev, curr) => prev + curr)
+
+  // Calculate the total storage for the full transaction
+  const totalStorageSize = rootStorageSize + nestedStorageSize
+
+  // Gas per byte ratio = Block Gas Limit / (Block Storage Limit (kb) * 1024 Bytes)
+  const ratio = 15_000_000 / (40 * 1024)
+
+  // Total gas cost for storage on moonbeam
+  // The ratio isn't an exact integer, so we round the result up
+  const moonbeamStorageCost = Math.ceil(totalStorageSize * ratio)
+
+  // Calculate the cost of the transaction according to Foundry without any storage usage
+  const baseGasLessStorageCost = Number(foundryGas) - 200 * totalStorageSize
```

Comment:
I get the idea, but where did 200 come from? For context, here's a [website](https://www.evm.codes/?fork=shanghai) that breaks down the cost of EVM opcodes, including SSTORE

Reply:
Good question, this is difficult to get exactly right because we're trying to subtract the exact cost of storing the data. SSTORE is a pretty complicated opcode. The gas cost is highly situational and depends on if there's data in the slot, if the slot is warm etc. The gas formula is designed to consider the processing power to complete the operation in addition to the intrinsic storage cost.

The 200 used here comes from the code deposit cost of 200 gas/byte. On the website you linked, you can see this for the create opcode if you go to the calculator and enter it 1 for the bottom row `Estimated deployed code size in bytes`. You'll see it adds 200 dynamic gas to the cost. You can also read about this [here](https://github.com/wolflo/evm-opcodes/blob/main/gas.md#a9-f-code-deposit-cost). Unfortunately, I couldn't find the exact EIP where this is defined. This cost is also applied to the create2 opcode.

I had seen documentation elsewhere that implied that the gas cost per byte for other storage methods was the same. However, after spending a while looking I wasn't able to find concrete documentation on that. After reading through the documentation on SSTORE, it doesn't seem like SSTORE has a similar type of direct cost per byte to what create and create2 have.

So, I think the safest thing to do would be to only subtract the 200 bytes for the storage size of contracts and not subtract anything for other storage mechanisms. That way we won't wildly overestimate the cost when deploying contracts, and we won't run the risk of subtracting too much because of other writes.